src/tools/rest-backend/authentication_application.cpp

```diff
@@ -0,0 +1,307 @@
+					     strlen (password.c_str ()), sha_out))
+		{
+			RootApp::setInternalServerError (response (), "Could not hash password. Please try again.", "AUTH_UNKNOWN_ERROR");
+			return;
+		}
+		if (u.getPasswordHash ().compare (std::string (reinterpret_cast<const char *> (sha_out))) != 0)
+		{
+			RootApp::setUnauthorized (response (), "The entered password is wrong.", "AUTH_INVALID_PASSWORD");
+			return;
+		}
+
+		/* authentication was successful, so lets build the token */
+		::jwt_t * jwt;
+
+		// reserve token memory and set algorithm
+		if (::jwt_new (&jwt) != 0 ||
+		    ::jwt_set_alg (jwt, JWT_ALG_HS256, reinterpret_cast<const unsigned char *> (ELEKTRA_REST_AUTHENTICATION_ENCRYPT_KEY),
+				   strlen (ELEKTRA_REST_AUTHENTICATION_ENCRYPT_KEY)) != 0)
+		{
+			RootApp::setInternalServerError (response (), "Something went wrong while creating the session token.",
+							 "AUTH_CREATE_TOKEN_ERROR");
+			::jwt_free (jwt); // free memory
+			return;		  // quit early due to error
+		}
+
+		// add claims
+		if (::jwt_add_grant (jwt, "issuer", ELEKTRA_REST_AUTHENTICATION_JWT_ISSUER) != 0 ||
+		    ::jwt_add_grant (jwt, "username", username.c_str ()) != 0 || ::jwt_add_grant_int (jwt, "rank", u.getRank ()) != 0 ||
+		    ::jwt_add_grant_int (jwt, "expires", std::time (NULL) + ELEKTRA_REST_AUTHENTICATION_JWT_EXPIRATION_TIME) != 0)
+		{
+			RootApp::setInternalServerError (response (), "Something went wrong while creating the session token.",
+							 "AUTH_CREATE_TOKEN_ERROR");
+			::jwt_free (jwt); // free memory
+			return;		  // quit early due to error
+		}
+
+		// generate token
+		std::string token (jwt_encode_str (jwt));
+		::jwt_free (jwt); // free memory
+
+		// write response
+		cppcms::json::value data;
+		data["token"] = token;
+		RootApp::setOk (response (), data, "application/json");
+	}
+	else if (request ().request_method () == "OPTIONS")
+	{
+		RootApp::setOk (response ());
+		return;
+	}
+	else
+	{
+		RootApp::setMethodNotAllowed (response ()); // send HTTP 405
+		return;
+	}
+}
+
+/**
+ * @brief helper method that allows to validate whether a request contains a valid
+ *		  session token or not. additionally the method can do checks on the authenticated
+ *		  users rank and username (i.e. check for owner of something).
+ * @note if a rank and a username are given, it is sufficient that one of both matches the
+ *       requirement in order for the validation to succeed.
+ * @param request a request
+ * @param response a response
+ * @param rank optionally the rank of a user can be checked
+ * @param orUser optionally the name of a user can be checked
+ * @return true if a authentication is present and the authenticated user matches all requirements,
+ *		   false otherwise
+ */
+bool AuthenticationApp::validateAuthentication (cppcms::http::request & request, cppcms::http::response & response, int rank,
+						std::string orUser)
+{
+#ifdef ELEKTRA_REST_NO_AUTH
+	return true;
```

Comment:
Not needed, will delete it. Early in development I thought it might be useful, but it isn't.